payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -345,42 +347,30 @@ public Payment apply(final PaymentModelDao paymentModelDao) {
                                               final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext tenantContext, final InternalTenantContext internalTenantContext) throws PaymentApiException {
         final PaymentPluginApi pluginApi = getPaymentPluginApi(pluginName);
 
-        return getEntityPagination(limit,
-                                   new SourcePaginationBuilder<PaymentTransactionInfoPlugin, PaymentApiException>() {
-                                       @Override
-                                       public Pagination<PaymentTransactionInfoPlugin> build() throws PaymentApiException {
-                                           try {
-                                               return pluginApi.searchPayments(searchKey, offset, limit, properties, tenantContext);
-                                           } catch (final PaymentPluginApiException e) {
-                                               throw new PaymentApiException(e, ErrorCode.PAYMENT_PLUGIN_SEARCH_PAYMENTS, pluginName, searchKey);
-                                           }
-                                       }
+        final Pagination<PaymentTransactionInfoPlugin> paymentTransactionInfoPlugins;
+        try {
+            paymentTransactionInfoPlugins = pluginApi.searchPayments(searchKey, offset, limit, properties, tenantContext);
```

Comment:
> I think the `offset` and `limit` are miss-leading here with the method name `searchPayments`. The client uses them to restrict the result payments size, but the plugin will use them on the transaction level.

True. That being said, the plugin could use it at the payment level instead, even if the search key matches the transactions (for 100 transactions matching, return the ones for the first `limit` payments).

> So it would be the case that, the client sends a limit of 100, but only get 1 payment in the page with 100 transactions embedded.

Typically though, a payment has 2-3 transactions (e.g. AUTH, CAPTURE, and maybe REFUND).

> Are we considering to have the plugins returning a more proper list instead of `PaymentTransactionInfoPlugin`s?

`PaymentPluginApi` changes aren't allowed anymore on 0.18.

> Just a thought, I'm happy it's not changed in this PR.

Were you able to confirm the patch solves your issues?

Reply:
> Were you able to confirm the patch solves your issues?
Sure, the patch has solved the issue that the method will return a null result when all returned `PaymentTransactionInfoPlugin`s in the `Pagination` belong to a single `Payment`.